airbyte-workers/src/main/java/io/airbyte/workers/DefaultReplicationWorker.java

```diff
@@ -95,21 +106,59 @@ public StandardSyncOutput run(StandardSyncInput syncInput, Path jobRoot) throws
         destination.start(destinationConfig, jobRoot);
         source.start(sourceConfig, jobRoot);
 
-        while (!cancelled.get() && !source.isFinished()) {
-          final Optional<AirbyteMessage> maybeMessage = source.attemptRead();
-          if (maybeMessage.isPresent()) {
-            final AirbyteMessage message = mapper.mapMessage(maybeMessage.get());
-
-            messageTracker.accept(message);
-            destination.accept(message);
+        final Thread sourceThread = new Thread(() -> {
```

Comment:
You should use a `newFixedThreadPool(1)` instead. It is generally error prone to use thread directly and it will make the `join` cleaner